2020/day3.clj

```diff
@@ -7,27 +7,39 @@
        (string/split-lines)))
 
 (defn next-location
-  [loc [right down]]
-  [(+ (first loc) right)
-   (+ (second loc) down)])
+  [[row col] [right down]]
+  [(+ row right)
+   (+ col down)])
 
 (defn locations-s
-  [height [right down]]
-  (take-while #(< (second %) height)
-              (map #(next-location [0 0] [(* right %) (* down %)])
-                   (range))))
+  [slope]
+  (let [right (first slope)
+        down  (second slope)]
+    (map (fn [n]
+           (next-location [0 0] [(* right n) (* down n)]))
+         (range))))
+
+(defn valid-locations
+  [height locations]
+  (->> locations
+       (take-while (fn [[row col]] (< col height)))))
+
+(defn char-at-location
+  [grid [row col]]
+  (let [x (grid col)
+        y (mod row (count x))]
+    (nth x y)))
 
 (defn tree?
   [ch]
   (= ch \#))
 
 (defn count-trees-on-slope
-  [grid [right down]]
-  (let [height          (count grid)
-        valid-locations (locations-s height [right down])]
-    (->> valid-locations
-         (map #(get-in grid [(second %) (mod (first %)
-                                             (count (first grid)))]))
+  [grid slope]
+  (let [height    (count grid)
+        locations (locations-s slope)]
+    (->> (valid-locations height locations)
```

Comment:
Did not address the previous comment. Stop capping your sequence length here

Reply:
> Did not address the previous comment. Stop capping your sequence length here

For the previous comment "(count grid) doesn't mean anything on it's own...". I've naming height for "(count grid)". Why we need to stop capping the sequence length here?

Reply:
There is nothing 'valid' about those locations until it's combined with a grid... but this is before it's combined with the grid.

Reply:
> There is nothing 'valid' about those locations until it's combined with a grid... but this is before it's combined with the grid.

For making sure the sequence length is capped only when it's combined with the grid information.

Can we remove the "valid-locations" function, then directly use the "take-while" in the threading to solve it?

Reply:
Yes, there shouldn't be a `valid-fns` . You want to combine.... and only _after_ that take valid inputs. You'll need to adjust your locations fn to nil-pun (or do something else) on invalid inputs